src/Provider/DefaultSmsProvider.php

```diff
@@ -44,145 +51,125 @@ class DefaultSmsProvider implements SmsProviderInterface {
     $this->configFactory = $config_factory;
     $this->moduleHandler = $module_handler;
   }
 
   /**
    * {@inheritdoc}
    */
-  public function queue(SmsMessageEntityInterface &$sms_message) {
-    if (!$sms_message->getGateway()) {
-      $sms_message->setGateway($this->getDefaultGateway());
-    }
+  public function queue(SmsMessageEntityInterface $sms_message) {
+    $sms_messages = $this->dispatch('sms.message.preprocess', [$sms_message]);
+
+    /** @var SmsMessageEntityInterface[] $sms_messages */
+    foreach ($sms_messages as $gateway_id => &$sms_message) {
+      // Tag so 'sms.message.*process' are not dispatched again.
+      $sms_message->setOption('_no_dispatch_events', TRUE);
 
-    if ($sms_message->getGateway()->getSkipQueue()) {
-      switch ($sms_message->getDirection()) {
-        case Direction::INCOMING:
-          $this->incoming($sms_message);
-          return;
-        case Direction::OUTGOING:
-          $this->send($sms_message);
-          return;
+      if ($count = $sms_message->validate()->count()) {
+        throw new SmsException(sprintf('Can not queue SMS message because there are %s validation error(s).', $count));
       }
-      return;
-    }
 
-    if ($count = $sms_message->validate()->count()) {
-      throw new SmsException(sprintf('Can not queue SMS message because there are %s validation error(s).', $count));
-    }
+      if ($sms_message->getGateway()->getSkipQueue()) {
+        switch ($sms_message->getDirection()) {
+          case Direction::INCOMING:
+            $this->incoming($sms_message);
+            continue;
+          case Direction::OUTGOING:
+            $this->send($sms_message);
+            continue;
+        }
+        continue;
+      }
 
-    // Split messages to overcome gateway limits.
-    $max = $sms_message->getGateway()->getMaxRecipientsOutgoing();
-    foreach ($sms_message->chunkByRecipients($max) as $sms_message) {
       $sms_message->save();
     }
+
+    return $this->dispatch('sms.message.postprocess', $sms_messages);
   }
 
   /**
    * {@inheritdoc}
    */
   public function queueIn(SmsMessageInterface $sms_message) {
-    $this->plainMessageQueue($sms_message, Direction::INCOMING);
+    $sms_message = SmsMessage::convertFromSmsMessage($sms_message)
+      ->setDirection(Direction::INCOMING);
+    $this->queue($sms_message);
   }
 
   /**
    * {@inheritdoc}
    */
   public function queueOut(SmsMessageInterface $sms_message) {
-    $this->plainMessageQueue($sms_message, Direction::OUTGOING);
+    $sms_message = SmsMessage::convertFromSmsMessage($sms_message)
+      ->setDirection(Direction::OUTGOING);
+    $this->queue($sms_message);
   }
 
   /**
    * {@inheritdoc}
    */
   public function send(SmsMessageInterface $sms) {
-    if (!$sms->getGateway()) {
-      $sms->setGateway($this->getDefaultGateway());
-    }
+    $dispatch = !$sms->getOption('_no_dispatch_events');
+    $sms_messages = $this->dispatch('sms.message.preprocess', [$sms], $dispatch);
```

Comment:
We could use conditional assignment here (and one other place I think) rather than passing in the extra `$dispatch` parameter.